Index: linux/arch/mips/ath79/mach-tl-wdr3500.c
===================================================================
--- linux.orig/arch/mips/ath79/mach-tl-wdr3500.c	2017-04-05 12:29:33.363023001 +0800
+++ linux/arch/mips/ath79/mach-tl-wdr3500.c	2017-09-07 11:53:34.866876000 +0800
@@ -105,7 +105,7 @@
 
 static void __init wdr3500_setup(void)
 {
-	u8 *mac = (u8 *) KSEG1ADDR(0x1f01fc00);
+	//u8 *mac = (u8 *) KSEG1ADDR(0x1f01fc00);
 	u8 *art = (u8 *) KSEG1ADDR(0x1fff0000);
 	u8 tmpmac[ETH_ALEN];
 
@@ -116,10 +116,10 @@
 					ARRAY_SIZE(wdr3500_gpio_keys),
 					wdr3500_gpio_keys);
 
-	ath79_init_mac(tmpmac, mac, 0);
+	ath79_init_mac(tmpmac, art + WDR3500_WMAC_CALDATA_OFFSET+2, 0);
 	ath79_register_wmac(art + WDR3500_WMAC_CALDATA_OFFSET, tmpmac);
 
-	ath79_init_mac(tmpmac, mac, 1);
+	ath79_init_mac(tmpmac, art + WDR3500_PCIE_CALDATA_OFFSET+2, 0);
 	ap9x_pci_setup_wmac_led_pin(0, 0);
 	ap91_pci_init(art + WDR3500_PCIE_CALDATA_OFFSET, tmpmac);
 
@@ -128,7 +128,7 @@
 	ath79_register_mdio(1, 0x0);
 
 	/* LAN */
-	ath79_init_mac(ath79_eth1_data.mac_addr, mac, -1);
+	ath79_init_mac(ath79_eth1_data.mac_addr, art + WDR3500_WMAC_CALDATA_OFFSET+2, -1);
 
 	/* GMAC1 is connected to the internal switch */
 	ath79_eth1_data.phy_if_mode = PHY_INTERFACE_MODE_GMII;
@@ -136,7 +136,7 @@
 	ath79_register_eth(1);
 
 	/* WAN */
-	ath79_init_mac(ath79_eth0_data.mac_addr, mac, 2);
+	ath79_init_mac(ath79_eth0_data.mac_addr, art + WDR3500_WMAC_CALDATA_OFFSET+2, 1);
 
 	/* GMAC0 is connected to the PHY4 of the internal switch */
 	ath79_switch_data.phy4_mii_en = 1;
